thick-wrappers/helper/utility.py

```
@@ -0,0 +1,233 @@
+
+from .constant import (
+    DESCRIPTION, EXIF_MAKE, SOFTWARE_TITLE, STARTING_X, STARTING_Y,
+    TEAM_MEMBERS, Direction, ExifData, ResizeMode, Sizing
+)
+
+T = TypeVar("T", int, np.signedinteger[Any])
+
+EXIF_MODEL_PATTERN = re.compile(r"I(?P<width>\d*)P(?P<height>\d*)P")
+
+
+def pixels_to_binary(img: Image.Image) -> list[tuple[str, str, str]]:
+    """
+    Translates an image's pixels to binary.
+
+    :param img: A Pillow Image object.
+    :return: A list of the tuples containing
+    strings representing the bits of each pixel's
+    red, green, and blue values.
+    """
+    pixels = img.load()
+    width, height = img.size
+    pixellist = []
+
+    # Iterates through all of the picture's pixels, left to right then down
+    for y in range(height):
+        for x in range(width):
+            r, g, b = pixels[x, y][:N_PLANES]
+            r, g, b = (
+                str(bin(r))[2:].zfill(8),
+                str(bin(g))[2:].zfill(8),
+                str(bin(b))[2:].zfill(8),
+            )
+            pixellist.append((r, g, b))
+    return pixellist
+
+
+def clear_least_significant_bits(bits: T, n: int) -> T:
+    """
+    Replaces an integer's `n` least significant bits with zeroes.
+
+    :param bits: An int, including an array of a NumPy int type
+    :param n: Number of bits to clear
+    """
+    bits >>= n
+    return bits << n
+
+
+def shift_image_bits_asarray(
+    image_array: np.ndarray, direction: Direction, bit_amount: int
+) -> np.array:
+    """
+    Converts image to numpy array and performs a left or right shift.
+
+    :param image_array: Image numpy array value
+    :param direction: Left bit shit or Right bit shift
+    :param bit_amount: Amount of bits to shift
+    :return: Image data as array
+    """
+    match direction:
+        case Direction.LEFT:
+            return np.left_shift(image_array, bit_amount)
+        case Direction.RIGHT:
+            return np.right_shift(image_array, bit_amount)
```

Nice use of pattern matching. Perhaps you could raise an exception if someone passes one of the other two values for Direction.